src/Extensions/Diagnostics.HealthChecks/AbstractHealthCheck.cs

```diff
@@ -57,13 +58,23 @@ public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context
 			catch (Exception exception)
 			{
 				Logger.LogError(Tag.Create(), exception, "'{0}' check failed with exception", context.Registration.Name);
-				return HealthCheckResult.Unhealthy("HealthCheck failed", exception, Parameters.ReportData);
+				return new HealthCheckResult(context.Registration.FailureStatus, "HealthCheck failed", exception, Parameters.ReportData);
 			}
 		}
 
 		/// <summary>
 		/// Health check logic without error handling and timed scope wrapping
 		/// </summary>
 		protected abstract Task<HealthCheckResult> CheckHealthInternalAsync(HealthCheckContext context, CancellationToken token);
+
+		private static HealthCheckResult EnforceFailureStatus(HealthStatus failureStatus, HealthCheckResult result)
+		{
+			if (result.Status == HealthStatus.Healthy)
+			{
+				return result;
+			}
+
+			return new HealthCheckResult(failureStatus, result.Description, result.Exception, result.Data);
```

Comment:
I think we should change logic a bit here.
First of all health checks with Error failure status could still report Degraded if they like
Second, we should not create a new struct if the status already correct.

Reply:
No, I believe the intention behind health checks is to make them return binary yay-nay responses, but user can configure what kind of nay.
Second issue fixed in a follow-up PR.